src/server/frontend_wayland/wl_surface.cpp

```diff
@@ -66,6 +79,12 @@ mf::WlSurface::~WlSurface()
     *destroyed = true;
     if (auto session = get_session(client))
         session->destroy_buffer_stream(stream_id);
+    role->destroy();
+}
+
+bool mf::WlSurface::synchronized() const
```

Comment:
This is the WlSurfaceRole abstraction getting leaky :grin:

We probably don't want WlSurface to implement the all methods that all wl_surface roles allow. I don't have a better idea at the moment; this is just a thought to look out for.